### PR TITLE
Merge CUDAFuture into ivalue::Future

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57052 Merge CUDAFuture into ivalue::Future**
* #57051 Make CUDAFuture handle any kind of device type
* #57050 Use only generic helpers in CUDAFuture
* #57049 Introduce generic MultiStreamGuard
* #57048 Simplify CUDAMultiStreamGuard
* #57047 Add recordDataPtrOnStream to DeviceGuardImplInterface
* #57046 Add getStreamFromPool to DeviceGuardImplInterface

This PR caps a stack whose goal was to merge CUDAFuture into ivalue::Future. CUDAFuture used to be a subclass of ivalue::Future, which was already pretty good, but it meant that in several places we needed `#ifdef`s or registries in order to create the right type of class, which was annoying. We've made CUDAFuture device-agnostic, by using generic helpers, so that it doesn't depend on CUDA. Now all its code can be inserted into ivalue::Future.

This PR does this very naively, by copy-pasting CUDAFuture's code into the (previously empty) virtual methods of ivalue::Future. This helps ensure the correctness of this PR, as it's straightforward to see it behaves exactly like before. However we probably want to polish it a bit later to iron out so wrinkles.

Differential Revision: [D28036829](https://our.internmc.facebook.com/intern/diff/D28036829/)

### CODE REVIEW

File: aten/src/ATen/core/ivalue.cpp [hunk 1]
```diff
@@ -932,6 +932,34 @@ getClassConverter() {
   return classConverter;
 }
 
+// Needs to be in this .cpp file to access the full definition of PyObjectHolder
+std::vector<std::reference_wrapper<const at::DataPtr>> ivalue::Future::extractDataPtrs(
```
Comment (mrshenli) [inline aten/src/ATen/core/ivalue.cpp:936]: assuming no change in the implementation, except this is moved from `CUDAFuture`.

File: aten/src/ATen/core/ivalue_inl.h [hunk 2]
```diff
@@ -311,7 +316,19 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
   }
 
  public:
-  explicit Future(TypePtr type) : type_(type) {}
+  explicit Future(TypePtr type, std::vector<c10::Device> devices={})
+    : type_(std::move(type)),
```
Comment (mrshenli) [inline aten/src/ATen/core/ivalue_inl.h:320]: this should be 4 spaces indent?

File: aten/src/ATen/core/ivalue_inl.h [hunk 3]
```diff
@@ -529,10 +551,44 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
...
+  void preMarkCompletedHook(
       const at::IValue& value,
       c10::optional<std::vector<std::reference_wrapper<const at::DataPtr>>>
-          data_ptrs) {}
+          dataPtrs) {
+    if (impl_ == nullptr) {
+      return;
+    }
+
+    // Start by performing all steps that can throw, before setting any field.
```
Comment (mrshenli) [inline aten/src/ATen/core/ivalue_inl.h:562]: also assuming no change in the implementation below.

**Review by mrshenli — APPROVED** [review]
This is great! Looks like we don't have hard blockers for this stack. I am going to update and land #57085 after this one.

**Review by mrshenli — CHANGES_REQUESTED** [review]
There are internal test failures.

File: aten/src/ATen/core/ivalue.cpp (hunk 1 above)
Comment (lw) [reply aten/src/ATen/core/ivalue.cpp:936]: Yes I just copy-pasted.

**Review by mrshenli — APPROVED** [review]
stamp to unblock

**Pull request closed without merging**